File: sdk/src/fleet.ts
import { ACCOUNT_TYPES } from "./generated/accounts";
import { encodePk } from "./bytes";

/**
 * Fleet readout from live account offsets.
 * TOTAL HULL / SHIELD are NOT on the fleet — client sums Game ship configs.
 */
export type FleetReadout = {
  fleetLabel: string;
  ownerProfile: string;
  subProfile: string;
  requiredCrew: number;
  passengerCapacity: number;
  respawnTimeWithoutFee: number;
  crewCount: number;
  rentedCrew: number;
  ap: number;
  fuelId: number;
  fuelAmount: bigint;
  ammoId: number;
  ammoAmount: bigint;
};

function off(name: string): number {
  const f = ACCOUNT_TYPES.Fleet.fields.find((x) => x.name === name);
  if (!f) throw new Error(`Fleet field ${name} missing`);
  return f.offset;
}

export function decodeFleetReadout(data: Buffer): FleetReadout {
  const labelAt = off("fleetLabel");
  return {
    fleetLabel: data.subarray(labelAt, labelAt + 64).toString("utf8").replace(/\0+$/, ""),
    ownerProfile: encodePk(data.subarray(off("ownerProfile"), off("ownerProfile") + 32)),
    subProfile: encodePk(data.subarray(off("subProfile"), off("subProfile") + 32)),
    requiredCrew: data.readUInt16LE(off("requiredCrew")),
    passengerCapacity: data.readUInt16LE(off("passengerCapacity")),
    respawnTimeWithoutFee: data.readUInt16LE(off("respawnTimeWithoutFee")),
    crewCount: data.readUInt16LE(off("crewCount")),
    rentedCrew: data.readUInt16LE(off("rentedCrew")),
    ap: data.readUInt32LE(off("ap")),
    fuelId: data.readUInt16LE(off("fuelTank.cargoId")),
    fuelAmount: data.readBigUInt64LE(off("fuelTank.amount")),
    ammoId: data.readUInt16LE(off("ammoBank.cargoId")),
    ammoAmount: data.readBigUInt64LE(off("ammoBank.amount")),
  };
}

File: sdk-rs/src/fleet.rs
use crate::bytes::encode_pk;
use crate::catalog::catalog;

/// Fleet readout from live account offsets. TOTAL HULL is not on-chain.
#[derive(Debug, Clone)]
pub struct FleetReadout {
    pub fleet_label: String,
    pub owner_profile: String,
    pub sub_profile: String,
    pub required_crew: u16,
    pub passenger_capacity: u16,
    pub respawn_time_without_fee: u16,
    pub crew_count: u16,
    pub rented_crew: u16,
    pub ap: u32,
    pub fuel_id: u16,
    pub fuel_amount: u64,
    pub ammo_id: u16,
    pub ammo_amount: u64,
}

fn u16_at(d: &[u8], o: usize) -> u16 {
    u16::from_le_bytes(d[o..o + 2].try_into().unwrap())
}
fn u32_at(d: &[u8], o: usize) -> u32 {
    u32::from_le_bytes(d[o..o + 4].try_into().unwrap())
}
fn u64_at(d: &[u8], o: usize) -> u64 {
    u64::from_le_bytes(d[o..o + 8].try_into().unwrap())
}

fn off(name: &str) -> Result<usize, String> {
    let fleet = catalog()
        .account_type("Fleet")
        .ok_or("Fleet type missing")?;
    let f = fleet
        .fields
        .iter()
        .find(|f| f.name == name)
        .ok_or_else(|| format!("Fleet field {name} missing"))?;
    let o = f
        .offset
        .as_ref()
        .and_then(|v| v.as_u64())
        .ok_or_else(|| format!("Fleet field {name} has no offset"))?;
    Ok(o as usize)
}

pub fn decode_fleet_readout(data: &[u8]) -> Result<FleetReadout, String> {
    let label_at = off("fleetLabel")?;
    if data.len() < label_at + 64 {
        return Err("fleet too small".into());
    }
    let label = String::from_utf8_lossy(&data[label_at..label_at + 64])
        .trim_end_matches('\0')
        .to_string();
    let owner = off("ownerProfile")?;
    let sub = off("subProfile")?;
    Ok(FleetReadout {
        fleet_label: label,
        owner_profile: encode_pk(&data[owner..owner + 32]),
        sub_profile: encode_pk(&data[sub..sub + 32]),
        required_crew: u16_at(data, off("requiredCrew")?),
        passenger_capacity: u16_at(data, off("passengerCapacity")?),
        respawn_time_without_fee: u16_at(data, off("respawnTimeWithoutFee")?),
        crew_count: u16_at(data, off("crewCount")?),
        rented_crew: u16_at(data, off("rentedCrew")?),
        ap: u32_at(data, off("ap")?),
        fuel_id: u16_at(data, off("fuelTank.cargoId")?),
        fuel_amount: u64_at(data, off("fuelTank.amount")?),
        ammo_id: u16_at(data, off("ammoBank.cargoId")?),
        ammo_amount: u64_at(data, off("ammoBank.amount")?),
    })
}
